ecmascript/transforms/compat/src/es2017/async_to_generator.rs

```diff
@@ -970,55 +973,52 @@ struct AsyncFnBodyHandler {
 macro_rules! noop {
     ($name:ident, $T:path) => {
         /// Don't recurse into function.
-        fn $name(&mut self, f: $T) -> $T {
-            f
-        }
+        fn $name(&mut self, _f: &mut $T) {}
     };
 }
 
-/// TODO: VisitMut
-impl Fold for AsyncFnBodyHandler {
-    noop_fold_type!();
+impl VisitMut for AsyncFnBodyHandler {
+    noop_visit_mut_type!();
 
-    noop!(fold_fn_decl, FnDecl);
-    noop!(fold_fn_expr, FnExpr);
-    noop!(fold_constructor, Constructor);
-    noop!(fold_arrow_expr, ArrowExpr);
+    noop!(visit_mut_fn_expr, FnExpr);
+    noop!(visit_mut_constructor, Constructor);
+    noop!(visit_mut_arrow_expr, ArrowExpr);
+    noop!(visit_mut_fn_decl, FnDecl);
 
-    fn fold_expr(&mut self, expr: Expr) -> Expr {
-        let expr = expr.fold_children_with(self);
+    fn visit_mut_expr(&mut self, expr: &mut Expr) {
+        expr.visit_mut_children_with(self);
 
         match expr {
             Expr::Await(AwaitExpr { span, arg }) => {
                 if self.is_async_generator {
                     let callee = helper!(await_async_generator, "awaitAsyncGenerator");
                     let arg = Box::new(Expr::Call(CallExpr {
-                        span,
+                        span: *span,
                         callee,
-                        args: vec![arg.as_arg()],
+                        args: vec![arg.clone().as_arg()],
                         type_args: Default::default(),
                     }));
-                    Expr::Yield(YieldExpr {
-                        span,
+                    *expr = Expr::Yield(YieldExpr {
+                        span: *span,
                         delegate: false,
                         arg: Some(arg),
                     })
                 } else {
-                    Expr::Yield(YieldExpr {
-                        span,
+                    *expr = Expr::Yield(YieldExpr {
+                        span: *span,
                         delegate: false,
-                        arg: Some(arg),
+                        arg: Some(arg.clone()),
```

Comment:
Can you change this to `arg.take()`?

Reply:
Sure thing, thanks for the pointer. Updated.